Improved Features for 'Add an interface to mark the solution as invalid / out-of-bounds'

MengnanLi91 opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
The feature is requested to determine whether the solution is valid at the time of convergence ref #22814. Several improvements have been done according to the discussion under PR #22884. This capability will also fill the need of PR #14996 

## Design
<!--A concise description (design) of the enhancement.-->
A registration system is used to store the sections needed to mark as invalid/out-of-bounds. The occurrence of solution invalidity will be tracked using unique IDs. A summary table of solution invalid warnings will be generated at the end of each time step including the number of warnings in the latest iteration, single time iteration and the whole calculation. The solution will not be allowed if the solution invalid warning is detected at the time of convergence. 

The summary table of solution invalid warnings can be silenced by setting `Problem/allow_invalid_solution=true`. Then the converged solution will be allowed even there are still solution invalid warnings in the latest iteration, but a message will be generated in the end of the calculation as a reminder.

The solution invalid warning can also be printed out immediately after it is detected by setting `Problem/immediately_print_invalid_solution=true`.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->
The new feature can state the converged solution is invalid and give a summary table of solution invalid warnings for each time step.

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

loganharbour committed:
Generalize PerfGraphRegistry for use in mooseDoOnceOnFlag

refs #17896

MengnanLi91 committed:
Solution Invalid refs #22814

MengnanLi91 committed:
Solution Invalid refs #22814

MengnanLi91 committed:
Solution Invalid refs #22814

MengnanLi91 committed:
Solution Invalid refs #22814

MengnanLi91 committed:
add Forward declarations

MengnanLi91 committed:
add Forward declarations

MengnanLi91 committed:
add Forward declarations

MengnanLi91 committed:
Merge #17897 to current version

MengnanLi91 committed:
Merge #17897 to current version

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

style change for Precheck test.

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

MengnanLi91 committed:
style change for Precheck test.

MengnanLi91 committed:
style change for Precheck test.

MengnanLi91 committed:
Fix reset mechanism #22884

MengnanLi91 committed:
Fix reset mechanism #22884

MengnanLi91 committed:
Fix reset mechanism #22884

MengnanLi91 committed:
style changes for precheck test

MengnanLi91 committed:
style changes for precheck test

MengnanLi91 committed:
Add doco for solution invalid and add newsletter blurb

MengnanLi91 committed:
Add doco for solution invalid and add newsletter blurb

MengnanLi91 committed:
Work in progress on invalid solution

MengnanLi91 committed:
Work in progress on invalid solution

working on making test case

update test case(not done yet)

MengnanLi91 committed:
Work in progress on invalid solution

MengnanLi91 committed:
working on making test case

MengnanLi91 committed:
working on making test case

MengnanLi91 committed:
update test case(not done yet)

MengnanLi91 committed:
update test case(not done yet)

MengnanLi91 committed:
Add SolutionInvalidityRegistry

MengnanLi91 committed:
Add SolutionInvalidityRegistry

MengnanLi91 committed:
Add SolutionInvalidityRegistry

MengnanLi91 committed:
Add SolutionInvalidity Summary Table

MengnanLi91 committed:
Add SolutionInvalidity Summary Table

MengnanLi91 committed:
Add SolutionInvalidity Summary Table

MengnanLi91 committed:
Add SolutionInvalidity Summary Table

MengnanLi91 committed:
Add SolutionInvalidity descriptions

MengnanLi91 committed:
Update summary table and add test

MengnanLi91 committed:
Update summary table and add test

MengnanLi91 committed:
Update summary table and add test

MengnanLi91 committed:
parent a76314478b3b240602c7f3f611d9f53ab55d79ae
author MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1671649416 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672872012 -0700

parent a76314478b3b240602c7f3f611d9f53ab55d79ae
author MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1671649416 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672869399 -0700

parent a76314478b3b240602c7f3f611d9f53ab55d79ae
author MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1671649416 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672869270 -0700

parent a76314478b3b240602c7f3f611d9f53ab55d79ae
author MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1671649416 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672868931 -0700

Update summary table and add test

MengnanLi91 committed:
Update summary table and add test

MengnanLi91 committed:
Update SolutionInvalidity.md

MengnanLi91 committed:
update output method

Update solution_invalid.i

MengnanLi91 committed:
update output method

MengnanLi91 committed:
Update solution_invalid.i

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1274668/) on 1741a3f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23005/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23005/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format af6972966d15adebd660174ea2dded977d12c3e5`

MengnanLi91 committed:
update format

MengnanLi91 committed:
Update SolutionInvalidity.md

MengnanLi91 committed:
update SolutionInvalidity

Update SolutionInvalidity.C

Update SolutionInvalidity.C

MengnanLi91 committed:
update SolutionInvalidity

loganharbour commented on framework/src/utils/PerfGraph.C:
```diff
@@ -487,7 +487,7 @@ dataStore(std::ostream & stream, PerfGraph & perf_graph, void *)
   // We need to store the registry id -> section info map so that we can add
   // registered sections that may not be added yet during recover
-  dataStore(stream, perf_graph._perf_graph_registry._id_to_section_info, nullptr);
+  //  dataStore(stream, perf_graph._perf_graph_registry._id_to_section_info, nullptr);
```
Reminder for me to look at this

MengnanLi91 committed:
Update SolutionInvalidity.md

MengnanLi91 committed:
Update SolutionInvalidity.md

MengnanLi91 committed:
Update SolutionInvalidity.C

MengnanLi91 committed:
Update RunParallel.py

MengnanLi91 committed:
Test console error

MengnanLi91 committed:
Test console error

MengnanLi91 committed:
update console printout

MengnanLi91 committed:
update console printout

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1276030/) on cf18cb4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23005/site)

This comment will be updated on new commits.

loganharbour committed:
parent 16ecc007b4f053c58d817fb3d5b09dee3720c606
author Logan Harbour <loganhharbour@gmail.com> 1671745211 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672868105 -0700

parent 16ecc007b4f053c58d817fb3d5b09dee3720c606
author Logan Harbour <loganhharbour@gmail.com> 1671745211 -0700
committer MengnanLi91 <118846840+MengnanLi91@users.noreply.github.com> 1672867926 -0700

Get PerfGraph dataStore/Load working again

style change for Precheck test.

MengnanLi91 committed:
Update SolutionInvalidInterface.h

MengnanLi91 committed:
Update RunParallel.py

MengnanLi91 committed:
Update tests

Update tests

Update tests

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1276817/) on 2db8477 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>631939</th>
      <th colspan=4>#23005 2db847</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.51%</td>
      <td>84.53%</td>
      <td>+0.01%</td>
      <td>96.05%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>81498</td>
      <td>81593</td>
      <td>+95</td>
      <td>146</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14933</td>
      <td>14938</td>
      <td>+5</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23005/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23005/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23005/coverage/xfem)

</details>



This comment will be updated on new commits.

tophmatthews commented:
This should close #14996 as well...or maybe already has?

MengnanLi91 commented:
> #14996

Yes, I think this capability will fill the need of #14996 as well.

MengnanLi91 committed:
expand threshold from number to vector

MengnanLi91 committed:
fix moose test threads

MengnanLi91 committed:
back to number for testing

MengnanLi91 commented:
@loganharbour If you have time, could you take a look at this failed test--Docker Test moose test threads?

loganharbour committed:
Add mutex to setSolutionInvalid()

MengnanLi91 committed:
Update SolutionInvalidity.C

loganharbour commented on framework/src/materials/MaterialBase.C:
```diff
@@ -92,7 +92,8 @@ MaterialBase::MaterialBase(const InputParameters & parameters)
     GeometricSearchInterface(this),
     FunctorInterface(this),
-    SolutionInvalidInterface(this),
+    SolutionInvalidInterface(
+        getMooseApp(), *parameters.getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
```
We'll change this to pass in the MooseObject instead

loganharbour commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -260,6 +260,11 @@ FEProblemBase::validParams()
+  params.addParam<bool>(
+      "immediately_print_invalid_solution",
+      false,
+      "Set to true to immediately print out the invalid solution warning during the calculation");
```
```suggestion
      "Whether or not to report invalid solution warnings at the time the warning is produced instead of after the calculation");
```

loganharbour commented on framework/src/systems/NonlinearSystem.C:
```diff
@@ -208,6 +209,23 @@ NonlinearSystem::solve()
+  _app.solutionInvalidity().solutionInvalidAccumulation();
+
+  // output the occurence of solution invalid in a summarry table
+  if (!_fe_problem.allowInvalidSolution() && _app.solutionInvalidity().solutionInvalid())
+  {
+    _app.solutionInvalidity().print(_console);
+  }
+  else if (_fe_problem.allowInvalidSolution() && _app.solutionInvalidity().solutionInvalid())
+  {
+    mooseWarning("The Solution Invalidity warnings are detected but silenced! "
+                 "Use Problem/allow_invalid_solution=false to activate ");
+  }
```
```suggestion
  if (!_fe_problem.allowInvalidSolution() && _app.solutionInvalidity().solutionInvalid())
    _app.solutionInvalidity().print(_console);
  else if (_fe_problem.allowInvalidSolution() && _app.solutionInvalidity().solutionInvalid())
    mooseWarning("An invalid solution has been detected but is silenced!\n"
                 "Set Problem/allow_invalid_solution=false to activate.");
```

loganharbour commented on framework/src/utils/SolutionInvalidity.C:
```diff
@@ -0,0 +1,147 @@
+}
+
+SolutionInvalidity::~SolutionInvalidity() {}
+/// Count solution invalid occurrences for each solution id
```
All of the comments in these methods should be defined in the header file, not here

loganharbour commented on framework/src/utils/SolutionInvalidity.C:
```diff
@@ -0,0 +1,147 @@
+bool
+SolutionInvalidity::solutionInvalid() const
+{
+  unsigned int sum_of_elems = 0;
+
+  std::for_each(_solution_invalid_counts.begin(),
+                _solution_invalid_counts.end(),
+                [&](int n) { sum_of_elems += n; });
+  if (sum_of_elems < 1)
+  {
+    return false;
+  }
+  else
+  {
+    return true;
+  }
```
```suggestion
  for (const auto count : _solution_invalid_counts)
    if (count)
      return true;
  return false;
```

loganharbour commented on framework/src/utils/SolutionInvalidity.C:
```diff
@@ -0,0 +1,147 @@
+  }
+}
+
+/// Reset the number of solution invalid occurrences back to zero
+void
+SolutionInvalidity::resetSolutionInvalid()
+{
+  std::fill(_solution_invalid_counts.begin(), _solution_invalid_counts.end(), 0);
+}
+
+/// Reset the number of solution invalid occurrences back to zero
+void
+SolutionInvalidity::resetSolutionInvalidTimeIter()
+{
+  std::fill(_solution_invalid_timeiter_counts.begin(), _solution_invalid_timeiter_counts.end(), 0);
+}
+
+/// Pass the number of solution invalid occurrences from current iteration to comulative counters
+void
+SolutionInvalidity::solutionInvalidAccumulation()
+{
+  for (unsigned int i = 0; i < _solution_invalid_counts.size(); i++)
+  {
+    _solution_invalid_total_counts[i] += _solution_invalid_counts[i];
+    _solution_invalid_timeiter_counts[i] += _solution_invalid_counts[i];
+  }
+}
```
All if this will be simplified when we combine it into a single vector.

MengnanLi91 replied:
I finished the modification following the example you showed me

loganharbour commented on framework/src/utils/SolutionInvalidity.C:
```diff
@@ -0,0 +1,147 @@
+  {
+
+    // Now print out the sections that contain solution invalid info and occurences
+    for (unsigned int id = 0; id < _solution_invalid_counts.size(); id++)
+    {
+
+      vtable.addRow(_solution_invalidity_registry.sectionInfo(id)._name, // Section
+                    _solution_invalid_counts[id],          // Current Interation Warnings
+                    _solution_invalid_timeiter_counts[id], // Current Time Interation Warnings
+                    _solution_invalid_total_counts[id],    // Total Iternation Warnings
+                    _solution_invalidity_registry.sectionInfo(id)._message // Message
+      );
+    }
```
Only add to the table when the count > 0

loganharbour commented on framework/src/utils/SolutionInvalidityRegistry.C:
```diff
@@ -0,0 +1,51 @@
+SolutionInvalidityRegistry::SolutionInvalidityRegistry()
+  : GeneralRegistry<std::string, SolutionInvaliditySectionInfo>("SolutionInvalidityRegistry")
+{
+  // Reserve space so that re-allocation doesn't need to happen much
+  // This does not take much memory and, for most cases, will keep a single
+  // reallocation from happening
+  reserve(5000);
+}
```
```suggestion
```

Shouldn't be as necessary for this registry

friedmud commented on framework/doc/content/source/outputs/SolutionInvalidity.md:
````diff
@@ -0,0 +1,42 @@
+
+```
+Solution Invalid Warnings:
+-------------------------------------------------------------------------------------------------------
+|     Section     | Current | Timestep | Total |                        Message                       |
+-------------------------------------------------------------------------------------------------------
+| NonsafeMaterial |      16 |       64 |    64 | The diffusivity is greater than the threshold value! |
+-------------------------------------------------------------------------------------------------------
````
I think so - at the end of a simulation it would be nice to know how many total happened.

friedmud commented on framework/include/interfaces/SolutionInvalidInterface.h:
```diff
@@ -11,25 +11,41 @@
-  FEProblemBase & _si_fe_problem;
+  void setSolutionInvalid(SolutionID _solution_id);
+
+  /**
+   * Get the SolutionInvalidity
+   */
+  SolutionInvalidity & solutionInvalidity();
```
Probably for pulling into a Postprocessor?

MengnanLi91 replied:
This is deleted since the current version is not used Postprocessor. I'll bring it back if we want live print

friedmud commented:
@loganharbour : thanks for the in-depth review.  These are all good modifications!

MengnanLi91 committed:
Update SolutionInvalidity.C

MengnanLi91 committed:
Address reviewer's comments

MengnanLi91 committed:
Update SolutionInvalidInterface.h

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1284841/) on d983033 : invalidated by @MengnanLi91

MengnanLi91 committed:
simplified three counters into one

MengnanLi91 commented on framework/doc/content/source/outputs/SolutionInvalidity.md:
```diff
@@ -0,0 +1,42 @@
+  if (material_property > _threshold[0] && material_property < _threshold[1])
+  {
+    setSolutionInvalid(solution_id);
+  }
```
Now is like
`  if (_input_diffusivity > _threshold)
    flagInvalidSolution("The diffusivity is greater than the threshold value!");`

MengnanLi91 commented on framework/include/utils/SolutionInvalidity.h:
```diff
@@ -0,0 +1,101 @@
+  void printDebug(SolutionID _solution_id);
+
+  /// The SolutionInvalidityRegistry
+  SolutionInvalidityRegistry & _solution_invalidity_registry;
```
This function is used by SolutionInvalidInterface. I'll leave it public as we discussed.

MengnanLi91 commented on framework/include/utils/SolutionInvalidityRegistry.h:
```diff
@@ -0,0 +1,154 @@
+   * @param message The description of the solution invalid warning
+   * @return The ID of the section - use when counting solution invalid warning
+   */
+  SolutionID registerSection(const std::string & section_name, const std::string & message);
```
"register" can not be used for a class member which gives an error `Invalid storage class for a class member`, so the function is renamed as "registerInvalidObject"

MengnanLi91 committed:
Address reviewer's comment

MengnanLi91 committed:
Update SolutionInvalidity.md

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1287134/) on 83e7f2e : invalidated by @MengnanLi91

loganharbour commented on framework/src/utils/SolutionInvalidityRegistry.C:
```diff
@@ -0,0 +1,41 @@
+{
+}
+
+InvalidSolutionID
+SolutionInvalidityRegistry::registerInvalidObject(const std::string & object_name,
+                                                  const std::string & message)
+{
+  return actuallyRegisterSection(object_name, message);
+}
+
+InvalidSolutionID
+SolutionInvalidityRegistry::actuallyRegisterSection(const std::string & object_name,
+                                                    const std::string & message)
+{
+  const auto create_item = [&object_name, &message](const std::size_t id)
+  { return SolutionInvalidityInfo(id, object_name, message); };
+  return registerItem(object_name, create_item);
+}
```
Let's get rid of `actuallyRegisterSection` and move all of its code into `registerInvalidObject`, and then rename `registerInvalidObject` into `registerInvalidity`

loganharbour reviewed (changes requested):
Most of this is minor nitpicking. We're good to go after this! Excited to get it in!

MengnanLi91 committed:
Address reviewer's comments

MengnanLi91 committed:
Update SolutionInvalidity.md

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1292198/) on 733676b : invalidated by @MengnanLi91

loganharbour commented on framework/src/utils/SolutionInvalidity.C:
```diff
@@ -0,0 +1,121 @@
+SolutionInvalidity::FullTable
+SolutionInvalidity::summaryTable() const
+{
+  FullTable vtable({"Section", "Current", "Timestep", "Total", "Message"}, 4);
```
```suggestion
  FullTable vtable({"Object", "Current", "Timestep", "Total", "Message"}, 4);
```

loganharbour commented on framework/src/utils/SolutionInvalidity.C:
```diff
@@ -0,0 +1,121 @@
+  FullTable vtable({"Section", "Current", "Timestep", "Total", "Message"}, 4);
+
+  vtable.setColumnFormat({
+      VariadicTableColumnFormat::AUTO, // Section Name
```
```suggestion
      VariadicTableColumnFormat::AUTO, // Object Name
```

loganharbour commented on framework/src/utils/SolutionInvalidity.C:
```diff
@@ -0,0 +1,121 @@
+  });
+
+  vtable.setColumnPrecision({
+      1, // Section Name
```
```suggestion
      1, // Object Name
```

loganharbour commented on framework/src/utils/SolutionInvalidity.C:
```diff
@@ -0,0 +1,121 @@
+
+  vtable.setColumnFormat({
+      VariadicTableColumnFormat::AUTO, // Section Name
+      VariadicTableColumnFormat::AUTO, // Current Interation Warnings
```
Find a replace `Interation` -> `Iteration` this happens a few times

loganharbour commented on framework/src/utils/SolutionInvalidity.C:
```diff
@@ -0,0 +1,121 @@
+    if (entry.counts > 0)
+    {
+      const auto & info = _solution_invalidity_registry.item(id);
+      vtable.addRow(info._name,            // Section
```
```suggestion
      vtable.addRow(info._name,            // Name
```

loganharbour commented on framework/doc/content/source/problems/SolutionInvalidity.md:
````diff
@@ -0,0 +1,47 @@
+
+If any solution invalidity is detected during the solve, a summary table of solution invalid warnings will be generated and reported at the end of each time step as below:
+
+```
+Solution Invalid Warnings:
+-------------------------------------------------------------------------------------------------------
+|     Section     | Current | Timestep | Total |                        Message                       |
+-------------------------------------------------------------------------------------------------------
+| NonsafeMaterial |      16 |       64 |    64 | The diffusivity is greater than the threshold value! |
+-------------------------------------------------------------------------------------------------------
+
+```
````
Rebuild this table now that section -> object

loganharbour commented on framework/doc/content/source/problems/SolutionInvalidity.md:
```diff
@@ -0,0 +1,47 @@
+
+The Solution Invalid Warning can also be printed out immediately after it is detected by setting [!param](/Problem/FEProblem/immediately_print_invalid_solution) to `true`.
+
+The two functions used internally in the 'SolutionInvalidInterface' for registeration are shown below:
+
+`flagInvalidSolutionInternal()` is an internal method used by the flagInvalidSolution macro that flags a solution as invalid with the given ID.
+
+`registerInvalidSolutionInternal()` is an internal method used by the flagInvalidSolution macro to statically register an invalid solution case.
+
+The code is showing below:
+
+!listing /framework/include/interfaces/SolutionInvalidInterface.h re=void flagInvalid.*\s*.*\;
```
Let's remove the content about internal methods - we don't need users to know these.

loganharbour commented on framework/include/utils/PerfGraphRegistry.h:
```diff
@@ -42,19 +43,11 @@ namespace internal
+  PerfGraphSectionInfo() = default;
   PerfGraphSectionInfo(const PerfID id,
-                       const std::string & name,
+                       const std::string name,
```
This was resolved but they are still not references

MengnanLi91 committed:
Address reviewer's comments

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1293757/) on eb1cede : invalidated by @MengnanLi91

MengnanLi91 committed:
Update SolutionInvalidity.md

MengnanLi91 commented:
@loganharbour My bad >.< I double checked locally and it has already updated a while ago, but this change didn't push to the remote branch for some reason.

MengnanLi91 committed:
Update PerfGraphRegistry.h
